File: Two Pointers & Sliding Window/threeSum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        // idea is to fix the first num, then reduce it to 2 sum
        // O(n^2)
        for (int i = 0; i + 2 < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i-1]) 
                continue; //dedup
            
            int j = i + 1, k = nums.length - 1; // 2 pointers
            int target = -nums[i];
            
            // reduce to 2 sum
            while (j < k) {
                if (nums[j] + nums[k] == target) {
                    res.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++; 
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) j++;  // skip same result
                    while (j < k && nums[k] == nums[k + 1]) k--;  // skip same result
                } 
                else if (nums[j] + nums[k] > target) 
                    k--;
                else 
                    j++;
            }
        }
        return res;
    }
}

File: Two Pointers & Sliding Window/threeSum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # 2 pointers, O(n^2) fix i and reduce to 2 sum
        # nums[j] + nums[k] = - nums[i],  0 <= i < j < k < n
        nums.sort()
        res, n, target = [], len(nums), 0
        if n < 3: return []
        
        for i in range(n - 2):
            if i and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, n - 1
            while j < k:
                s = nums[i] + nums[j] + nums[k]
                if s == target:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]: 
                        j += 1 # dedup j
                    while j < k and nums[k] == nums[k + 1]: 
                        k -= 1 # dedup k
                elif s < target:
                    j += 1
                else:
                    k -= 1
        return res 
        